### DISCUSSION TITLE
Is it safe to enable exec()?

### DISCUSSION THREAD

**Discussion by pat-githubber** [HomeITAdmin/nextcloud_geoblocker#100] [Q&A]
Hi,
I have zero knowledge about how the coding works with GeoBlocker. I was just researching the `exec` method, which is mentioned in the documentation with respect to [Geoiplookup](https://github.com/HomeITAdmin/nextcloud_geoblocker#geoiplookup):

> PHP must be allowed to use the exec() method

In a forum post, it says [this](https://stackoverflow.com/questions/1295346/is-php-exec-safe):

> It's only a security hole if you let your user enter parameters. I.E., you shouldn't do this:
> 
> ```
> // DO NOT DO THIS!
> exec("foo -bar=" . $_GET['bar']);
> ```
> 
> Instead, you should sanitize your parameters with [escapeshellarg](http://us3.php.net/manual/en/function.escapeshellarg.php).
> 
> ```
> // This is okay.  (Be sure foo.exe can handle unexpected input!)
> exec("foo -bar=" . escapeshellarg($_GET['bar']));
> ```

Another [blog](https://bobcares.com/blog/how-to-enable-exec-function-in-php-ini/) states that:

> ### Security risk in enabling exec function
> 
> PHP has certain dangerous functions that can make a server vulnerable to attacks. These functions include `system, passthru, shell_exec, exec`, etc.
> 
> Hence our [Support Engineers](https://bobcares.com/server-management-services/) always recommend our customers to disable these functions in the server. And these functions will be listed in the `disable_functions` directive in a `php.ini` file.
>
> By enabling exec function there is a possible security threat. Attacks may use this function to run programs in servers that can lead to data corruption and data loss.
> 
> Considering the security risks, we suggest customers enabling the exec function on a per-account basis. Thus, it gives better tracking of user activity.
> 

Again, I have know clue what that means regarding my NextCloudPi 24, latest PHP 8.1

What is your take on that?

Creates the enabling of `exec` a greater risk than just letting abuse-IPs try to log-in and fail?

Best,
Pat

**Comment by HomeITAdmin** [comment]
Hi Pat,

yes, there is a risk and that is the reason, why it is deactivated often.

The risk from the Geoblocker App is minimal, because it is not taking any user input when calling it. Just the IP address and a static string is used. (https://github.com/HomeITAdmin/nextcloud_geoblocker/blob/master/lib/LocalizationServices/GeoIPLookupCmdWrapper.php)
But you cannot be sure what the rest of Nextcloud is doing with that right and what other applications using the same PHP config is doing with it.
That is why this is mentioned as a disadvantage.
I think overall this localization service is not the best option from the three currently provided. But for admins that know what they are doing still a valid one and it has some advantages.

How much sense the Geoblocker App makes at all and if this is bigger the the risk is another question, see also my comment in #99.

Best

**Comment by pat-githubber** [comment]
Thanks @HomeITAdmin 🙏 

> not the best option from the three currently provided

What are the other options?

I am asking because my issue is that certain abusive IPs try to log in on my NC-Pi. There is a log in attempt every 15 min and I do not know how to block these IPs. 
I can only find **Brute Force Whitelist** [Apps/Settings](https://apps.nextcloud.com/apps/bruteforcesettings) for NC and any **Brute Force Blacklist** options and I am not experience enough to code such a blacklist (or at least I have not found an explanation online how to do this)

Furthermore, GeoBlocker does not always work because e.g. [this abusive IP](https://www.abuseipdb.com/check/137.184.179.53) from USA tries to log in even though I put USA on the blacklist in GeoBlocker 🤷‍♂️ 

The coolest thing would be to automatically block all IPs, which are listed on [abuseipdb.com](abuseipdb.com) or so.

**Reply by HomeITAdmin** [reply]
You can see the other options in the Readme here in Github.

This App has a clear target what it is doing and you seems to look for something different. Best thing would be of cause to block the IPs on network level. 

As written in the mentioned issue, the Geoblocker App has some limitations. So if this is really a direct attack against your service, it will not prevent a good attacker from reaching its target.

Nevertheless, if you block US it should work. It could be a config problem or the last time I checked the database of geoiplookup was pretty old, so maybe that is a problem. You could check, what is geoiplookup saying for that IP on your machine.
